### DISCUSSION TITLE
how did you manage to configure/create vlans?

### DISCUSSION THREAD

**Discussion by ji-podhead** [O-X-L/ansible-opnsense#113] [Q&A] [answered]
This is the only repo/apiwrapper/collection... that allows to create vlans.
I can enable disable NICs/physical interfaces via config.xml but i cant create vlans (maybe this works too, but i did something wrong).
Im not sure if this collection can enable nics.
If im using the api to get/set interfaces im just getting `...404...controller not found`. Im not sure if thats a bug....
I need to be able to activate nics because i wrote a collection that creates ovs bridges and vlan interfaces, so i can attach them to libvirt in runtime. i can even switch them out without rebooting.
I looked at your code, but to follow up i need to further get into it, so i need to debug it since its a lot of complex code  ;)
Anyway,

Grüße aus Berlin und danke für die Collection:

ji

**Comment by ji-podhead** [comment]


**Reply by superstes** [reply]
Yes - we only use the officially supported APIs.

See:
* Docs - https://opnsense.ansibleguy.net/en/latest/modules/interface.html
* API - https://docs.opnsense.org/development/api/core/interfaces.html

**Reply by superstes** [reply, marked as answer]
You can see all API calls with payload if you enable the `debug` mode: https://opnsense.ansibleguy.net/en/latest/usage/3_troubleshoot.html#verbose-output

**Reply by superstes** [reply]
Yeah - the OPNSense API is not the cleanest. Their focus still is the support for the WebUI.
Sometimes you even need different values for POST than you GET.

These modules should be able to en- and disable these interfaces.
But when I am remembering correctly - the main interfaces are not yet API enabled 🤔

**Comment by ji-podhead** [comment]


**Reply by superstes** [reply]
It's not designed to be used without Ansible as we heavily depend on the objects provided by Ansible.
You could - of course - run Ansible programmatically. There are some examples in the Ansible docs. But I have no experience with that use-case.

**Reply by superstes** [reply]
We could create a Fork to abstract the basic code for re-use. Could make sense. Some refactoring would be needed.

**Reply by superstes** [reply]
Would be nice if you could test the fork that does has the Ansible-dependency removed:
https://github.com/O-X-L/opnsense-api-client

```python3
from src.oxl_opnsense_client import Client
with Client(firewall='<IP>', port=443, credential_file='<FILE>', ssl_verify=False) as c:
    c.test()
    # True
    c.run_module('syslog', params={'target': '192.168.0.1', 'port': 5303})
    # {'error': None, 'result': {'changed': True, 'diff': {'after': {'uuid': None, 'rfc5424': False, 'enabled': True, 'target': '192.168.0.1', 'transport': 'udp4', 'facility': [], 'program': [], 'level': ['alert', 'crit', 'emerg', 'err', 'info', 'notice', 'warn'], 'certificate': '', 'port': 5303, 'description': ''}}}}
    c.run_module('syslog', params={'target': '192.168.0.1', 'port': 5303, 'state': 'absent'})
    # {'error': None, 'result': {'changed': True, 'diff': {'before': {'uuid': '7f3aba31-07ca-4cb9-b93d-dc442a5291c7', 'rfc5424': False, 'enabled': True, 'target': '192.168.0.1', 'transport': 'udp4', 'facility': [], 'program': [], 'level': ['alert', 'crit', 'emerg', 'err', 'info', 'notice', 'warn'], 'certificate': '', 'port': 5303, 'description': ''}}}}
    c.run_module('syslog', params={'target': '192.168.0.1', 'port': 5303, 'state': 'absent'})
    # {'error': None, 'result': {'changed': False, 'diff': {}}}
```

**Reply by ji-podhead** [reply]
> Would be nice if you could test the fork that does has the Ansible-dependency removed: https://github.com/O-X-L/opnsense-api-client

Yes sure, I'll do that later when I get back home!

**Reply by ji-podhead** [reply]
i opened a discussion in [ the fork ](https://github.com/O-X-L/opnsense-api-client/discussions/1)
i found out how to bring up interfaces and reconfigure vlans using the backend.
i wrote a function using paramiko:
```python
def reconfigure_vlans(helper):
    stdin, stdout, stderr = helper.ssh_client.exec_command('/usr/local/opnsense/scripts/interfaces/reconfigure_vlans.php')
    output = stdout.read().decode('utf-8')
    print(output)
    error = stderr.read().decode('utf-8')
    if error:
        print(f"Fehler: {error}")
```
ill remove the entire frontend api stuff from my code. if you want i can change my aliases so they align with your module variables, so people can use the backend stuff too

here is my repo: [opnsense-helper](https://github.com/the-pod-shop/opnsense-helper)
its acutally much faster since its just parsing a file
